src/docs/dev_tasks.md

```diff
@@ -225,65 +225,40 @@ shows some useful idioms for JVM tasks.
 
-Automatic target workdir caching works by assigning a results directory to
-each VersionedTarget (VT) of the InvalidationCheck yielded by `Task->invalidated`.
 A task operating on a given VT should place the resulting artifacts in the VT's
 `results_dir`. After exiting the `invalidated` context block, these artifacts
 will be automatically uploaded to the artifact cache.
 
 This interface for caching is disabled by default. To enable, override
 `Task->cache_target_dirs` to return True.
 
-**Manual caching**
+**VersionedTargetSets**
 
-Manual caching is much more complicated than automatic target workdir caching,
-and as such should only be used for non-standard usecases. Instead of placing
-artifacts in known target directories, artifacts may be placed anywhere -- although
-it is now the responsibility of the task developer to manually upload
-VT / artifact pairs to the cache. Here is a template for how manual caching
-would be implemented in the `execute` method:
+By default, Pants caching operates on (VT, artifacts) pairs. But certain tasks may need to write multiple targets and their artifacts together under a single cache key. For example, Ivy resolution, where the set of resolved 3rd party dependencies is a property of all targets taken together, not of each target individually.
 
-    def execute(self):
-      targets = self.context.targets()
-      with self.invalidated(targets) as invalidation_check:
+To implement caching for groupings of targets, you can override the `check_artifact_cache_for` method in your task to check for the collected `VersionedTarget`s as a `VersionedTargetSet`:
 
-        # for each VersionedTarget in invalidation_check.invalid_vts,
-        # run your task on the target and remember where the output files are
-        output_files = do_some_work()
+    def check_artifact_cache_for(self, invalidation_check):
+      return [VersionedTargetSet.from_versioned_targets(invalidation_check.all_vts)]
 
-        if self.artifact_cache_writes_enabled():
-          # build a list of (VersionedTarget, output_files) pairs
-          pairs = match_up(invalidation_check.invalid_vts, output_files)
+**Manual caching**
 
-          self.update_artifact_cache(pairs)
+Pants allows more fine grained cache management, although it then becomes the responsibility of the task developer to manually upload VT / artifact pairs to the cache. Here is a template for how manual caching might be implemented:
 
-The above implementation writes each target/artifact (key/val) pair to the cache
-independently of all other targets. However you might want to write multiple
-targets and their artifacts together under a single cache key. A good example of
-this is Ivy resolution, where the set of resolved 3rd party dependencies is a property
-of all targets taken together, not of each target individually.
+    def execute(self):
+      targets = self.context.targets()
```

Comment:
Would be good to add some imaginary `predicate` arg here, to demonstrate that you generally wouldn't do this for all targets in the context.